Fix possible UB in case of MEMORY_LIMIT_EXCEEDED during String deserialization

azat opened the pull request:
### Changelog category (leave one):
- Critical Bug Fix (crash, data loss, RBAC) or LOGICAL_ERROR

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Fix possible UB (crashes) in case of MEMORY_LIMIT_EXCEEDED during String deserialization

Previously it was possible to leave `offsets` and `data` out of sync if `data.resize_exact` fails with `MEMORY_LIMIT_EXCEEDED`, which eventually will lead to various UBs (i.e. `SIZES_OF_COLUMNS_DOESNT_MATCH` and `LOGICAL_ERROR`)

azat committed:
Fix possible UB in case of MEMORY_LIMIT_EXCEEDED during String deserialization

Previously it was possible to leave `offsets` and `data` out of sync if
`data.resize_exact` fails with `MEMORY_LIMIT_EXCEEDED`, which eventually
will lead to various UBs (i.e. `SIZES_OF_COLUMNS_DOESNT_MATCH` and
`LOGICAL_ERROR`)

alexey-milovidov commented:
What is the scenario when we use an incomplete String column after an exception is thrown?

alexey-milovidov reviewed (changes requested):
Needs a test.

You can make a test by gradually increasing the `max_memory_usage` value.

alexey-milovidov commented:
PTAL why everything failed in this PR.

azat commented:
>What is the scenario when we use an incomplete String column after an exception is thrown?

https://github.com/ClickHouse/ClickHouse/blob/c0b3821d629f431bd8bcc5df9d1b9aefe8a0d4a7/src/Dictionaries/HashedArrayDictionary.cpp#L512

azat committed:
Add a unit test for string deserialization UB

W/o the fix it always fails, i.e.:

    $ src/unit_tests_dbms --gtest_filter=StringSerialization.*
    Note: Google Test filter = StringSerialization.*
    [==========] Running 1 test from 1 test suite.
    [----------] Global test environment set-up.
    [----------] 1 test from StringSerialization
    [ RUN      ] StringSerialization.IncorrectStateAfterMemoryLimitExceeded
    /src/ch/clickhouse/src/DataTypes/Serializations/tests/gtest_string_serialization.cpp:79: Failure
    Expected equality of these values:
      result.getDataAt(result.size() - 1)
        Which is: ar\0foo
      "foobar"
        Which is: 0x555557ce25a3

    [  FAILED  ] StringSerialization.IncorrectStateAfterMemoryLimitExceeded (28 ms)
    [----------] 1 test from StringSerialization (28 ms total)

    [----------] Global test environment tear-down
    [==========] 1 test from 1 test suite ran. (30 ms total)
    [  PASSED  ] 0 tests.
    [  FAILED  ] 1 test, listed below:
    [  FAILED  ] StringSerialization.IncorrectStateAfterMemoryLimitExceeded

     1 FAILED TEST

azat commented:
> Needs a test.

@alexey-milovidov https://github.com/ClickHouse/ClickHouse/pull/85440/commits/c3e4a8130ebee61413ac0af71f6df7e5c54a2447

azat committed:
tests/gtest_filecache: fix overlaps of ThreadStatus and MainThreadStatus

azat commented:
>[Stress test (amd_ubsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=85440&sha=latest&name_0=PR&name_1=Stress+test+%28amd_ubsan%29)

- https://github.com/ClickHouse/ClickHouse/issues/81144

>[Integration tests (arm_binary, distributed plan, 2/4)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=85440&sha=latest&name_0=PR&name_1=Integration+tests+%28arm_binary%2C+distributed+plan%2C+2%2F4%29)

- `test_async_logger_metrics` - https://github.com/ClickHouse/ClickHouse/pull/85578
